zng/walk.go

```diff
@@ -0,0 +1,288 @@
+package zng
+
+import (
+	"github.com/brimsec/zq/zcode"
+)
+
+// A RecordVisitor is called for each value in a record encountered by
+// Walk. Upon visiting a container value, the boolean returned by the
+// RecordVisitor determines whether the values of that
+// container value will be individually visited.
+// If the visitor returns an error, the walk stops and that error will be
+// returned to the caller of Walk().
+type RecordVisitor func(typ Type, body zcode.Bytes) (bool, error)
+
+func walkRecord(typ *TypeRecord, body zcode.Bytes, rv RecordVisitor) error {
+	if body == nil {
+		return nil
+	}
+	it := zcode.Iter(body)
+	for _, col := range typ.Columns {
+		if it.Done() {
+			return &RecordTypeError{Name: col.Name, Type: col.Type.String(), Err: ErrMissingField}
+		}
+		body, container, err := it.Next()
+		if err != nil {
+			return err
+		}
+		switch t := AliasedType(col.Type).(type) {
+		case *TypeRecord:
+			if !container {
+				return &RecordTypeError{Name: col.Name, Type: col.Type.String(), Err: ErrNotContainer}
+			}
+			enter, err := rv(t, body)
+			if err != nil {
+				return err
+			}
+			if !enter {
+				continue
+			}
+			if err := walkRecord(t, body, rv); err != nil {
+				return err
+			}
+		case *TypeArray:
+			if !container {
+				return &RecordTypeError{Name: col.Name, Type: col.Type.String(), Err: ErrNotContainer}
+			}
+			enter, err := rv(t, body)
+			if err != nil {
+				return err
+			}
+			if !enter {
+				continue
+			}
+			if err := walkVector(t, body, rv); err != nil {
+				return err
+			}
+		case *TypeSet:
+			if !container {
+				return &RecordTypeError{Name: col.Name, Type: col.Type.String(), Err: ErrNotContainer}
+			}
+			enter, err := rv(t, body)
+			if err != nil {
+				return err
+			}
+			if !enter {
+				continue
+			}
+			if err := walkSet(t, body, rv); err != nil {
+				return err
+			}
+		case *TypeUnion:
+			if !container {
+				return &RecordTypeError{Name: col.Name, Type: col.Type.String(), Err: ErrNotContainer}
+			}
+			enter, err := rv(t, body)
+			if err != nil {
+				return err
+			}
+			if !enter {
+				continue
+			}
+			if err := walkUnion(t, body, rv); err != nil {
+				return err
+			}
```

Comment:
These four cases are identical except for the function called, so you can probably refactor this so you're not repeating yourself.

Reply:
Great point. In fact there's a some repetition throughout these record traversal funcs, as there was in the TypeCheck and TypeIndexer that this ressembles. The good news is this repetition now only occurs once.

To this specific instance, the issue is that each function called has a different signature. So you can't pass it to a factored out case function, as far as I can tell. Specifically, I'd like to write a handler like:

```go
process := func(t Type, walker func(Type, zcode.Bytes, RecordVisitor) error) (bool, error) {
..
}
```

and then call it as:
```go
		case *TypeRecord:
			if err := process(t, walkRecord); err != nil {
				return err
			}
```

but this is a type error. Am I missing a better way to refactor ? Fwiw, the above refactor would save about 15 lines, if it worked.